File: android/app/src/main/java/com/signal/app/features/signal/presentation/components/MapVisualizationLayer.kt
package com.signal.app.features.signal.presentation.components

import androidx.compose.animation.*
import androidx.compose.animation.core.*
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.google.android.gms.maps.model.*
import com.signal.app.features.signal.data.services.DensityPoint
import com.signal.app.features.signal.data.services.SignalCluster
import com.signal.app.features.signal.data.services.Hotspot
import com.signal.app.features.signal.presentation.state.MapViewMode
import kotlin.math.max
import kotlin.math.min

/**
 * 지도 시각화 레이어 (밀도 맵, 클러스터, 핫스팟)
 */
object MapVisualizationLayer {

    /**
     * 밀도 맵을 위한 히트맵 오버레이 생성
     */
    fun createDensityHeatmap(
        densityPoints: List<DensityPoint>,
        gridSize: Double = 0.01,
        baseColor: Color = Color.Blue
    ): Set<Polygon> {
        val polygons = mutableSetOf<Polygon>()

        densityPoints.forEachIndexed { index, point ->
            val opacity = calculateOpacity(point.density, point.weight)

            // 그리드 셀을 다각형으로 변환
            val polygon = Polygon.Builder()
                .addAll(createGridCell(point.latitude, point.longitude, gridSize))
                .fillColor(baseColor.copy(alpha = opacity).toArgb())
                .strokeColor(baseColor.copy(alpha = opacity * 0.7f).toArgb())
                .strokeWidth(1f)
                .clickable(false)
                .build()

            polygons.add(polygon)
        }

        return polygons
    }

    /**
     * 클러스터 마커 생성
     */
    fun createClusterMarkers(
        clusters: List<SignalCluster>,
        onClusterTap: (SignalCluster) -> Unit
    ): Set<Marker> {
        val markers = mutableSetOf<Marker>()

        clusters.forEach { cluster ->
            val marker = MarkerOptions()
                .position(LatLng(cluster.centerLat, cluster.centerLon))
                .icon(createClusterIcon(cluster.signalCount))
                .title("클러스터")
                .snippet("${cluster.signalCount}개 시그널")

            markers.add(marker)
        }

        return markers
    }

    /**
     * 핫스팟 마커 생성
     */
    fun createHotspotMarkers(
        hotspots: List<Hotspot>,
        onHotspotTap: (Hotspot) -> Unit
    ): Set<Marker> {
        val markers = mutableSetOf<Marker>()

        hotspots.forEach { hotspot ->
            val marker = MarkerOptions()
                .position(LatLng(hotspot.latitude, hotspot.longitude))
                .icon(BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_ORANGE))
                .title("핫스팟: ${hotspot.name}")
                .snippet("인기 지역")

            markers.add(marker)
        }

        return markers
    }

    /**
     * 다각형 검색 영역 표시
     */
    fun createPolygonSearchArea(
        polygonPoints: List<List<Double>>,
        fillColor: Color = Color.Blue,
        strokeColor: Color = Color.Blue
    ): Set<Polygon> {
        if (polygonPoints.size < 3) return emptySet()

        val points = polygonPoints.map { point ->
            LatLng(point[1], point[0]) // [lon, lat] -> LatLng
        }

        val polygon = Polygon.Builder()
            .addAll(points)
            .fillColor(fillColor.copy(alpha = 0.2f).toArgb())
            .strokeColor(strokeColor.toArgb())
            .strokeWidth(2f)
            .clickable(false)
            .build()

        return setOf(polygon)
    }

    /**
     * 경로 검색 영역 표시
     */
    fun createRouteSearchArea(
        startLat: Double,
        startLon: Double,
        endLat: Double,
        endLon: Double,
        bufferWidth: Double = 1000.0,
        color: Color = Color.Green
    ): Set<Polyline> {
        val polyline = PolylineOptions()
            .add(LatLng(startLat, startLon))
            .add(LatLng(endLat, endLon))
            .color(color.toArgb())
            .width(5f)
            .pattern(listOf(Dash(20f), Gap(10f)))

        return setOf(polyline)
    }

    /**
     * 반경 검색 영역 표시
     */
    fun createRadiusSearchArea(
        centerLat: Double,
        centerLon: Double,
        radius: Double,
        fillColor: Color = Color.Blue,
        strokeColor: Color = Color.Blue
    ): Set<Circle> {
        val circle = CircleOptions()
            .center(LatLng(centerLat, centerLon))
            .radius(radius)
            .fillColor(fillColor.copy(alpha = 0.1f).toArgb())
            .strokeColor(strokeColor.toArgb())
            .strokeWidth(2f)

        return setOf(circle)
    }

    // Private Helper Methods

    private fun calculateOpacity(density: Int, weight: Double): Float {
        // 밀도와 가중치를 기반으로 투명도 계산
        val normalizedDensity = min(density / 10.0, 1.0)
        val normalizedWeight = min(weight / 5.0, 1.0)
        return max(normalizedDensity * normalizedWeight, 0.1).toFloat()
    }

    private fun createGridCell(
        centerLat: Double,
        centerLon: Double,
        gridSize: Double
    ): List<LatLng> {
        val halfGrid = gridSize / 2
        return listOf(
            LatLng(centerLat - halfGrid, centerLon - halfGrid),
            LatLng(centerLat - halfGrid, centerLon + halfGrid),
            LatLng(centerLat + halfGrid, centerLon + halfGrid),
            LatLng(centerLat + halfGrid, centerLon - halfGrid)
        )
    }

    private fun createClusterIcon(count: Int): BitmapDescriptor {
        // 클러스터 크기에 따른 아이콘 선택
        return when {
            count < 5 -> BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_BLUE)
            count < 10 -> BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_GREEN)
            count < 20 -> BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_YELLOW)
            else -> BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_RED)
        }
    }
}

/**
 * 실시간 애니메이션을 위한 커스텀 컴포저블
 */
@Composable
fun AnimatedMapOverlay(
    densityPoints: List<DensityPoint>,
    clusters: List<SignalCluster>,
    viewMode: MapViewMode,
    onClusterTap: (SignalCluster) -> Unit,
    modifier: Modifier = Modifier
) {
    val pulseAnimation = rememberInfiniteTransition()
    val pulseScale by pulseAnimation.animateFloat(
        initialValue = 0.8f,
        targetValue = 1.2f,
        animationSpec = infiniteRepeatable(
            animation = tween(2000),
            repeatMode = RepeatMode.Reverse
        )
    )

    var previousViewMode by remember { mutableStateOf(viewMode) }
    val fadeAnimation = remember { Animatable(1f) }

    LaunchedEffect(viewMode) {
        if (previousViewMode != viewMode) {
            fadeAnimation.animateTo(0f, animationSpec = tween(250))
            fadeAnimation.animateTo(1f, animationSpec = tween(250))
            previousViewMode = viewMode
        }
    }

    Box(
        modifier = modifier
            .fillMaxSize()
            .graphicsLayer(alpha = fadeAnimation.value)
    ) {
        when (viewMode) {
            MapViewMode.DENSITY -> DensityOverlay(
                densityPoints = densityPoints,
                pulseScale = pulseScale
            )
            MapViewMode.CLUSTER -> ClusterOverlay(
                clusters = clusters,
                onClusterTap = onClusterTap
            )
            else -> { /* 일반 모드 - 오버레이 없음 */ }
        }
    }
}

@Composable
private fun DensityOverlay(
    densityPoints: List<DensityPoint>,
    pulseScale: Float,
    modifier: Modifier = Modifier
) {
    Canvas(
        modifier = modifier.fillMaxSize()
    ) {
        densityPoints.forEach { point ->
            val opacity = calculateOpacity(point.density, point.weight)
            val radius = 20.dp.toPx() * pulseScale

            drawCircle(
                color = Color.Red.copy(alpha = opacity),
                radius = radius,
                center = Offset(
                    x = point.longitude.toFloat() * size.width, // 실제 좌표 변환 필요
                    y = point.latitude.toFloat() * size.height
                )
            )
        }
    }
}

@Composable
private fun ClusterOverlay(
    clusters: List<SignalCluster>,
    onClusterTap: (SignalCluster) -> Unit,
    modifier: Modifier = Modifier
) {
    Box(modifier = modifier.fillMaxSize()) {
        clusters.forEach { cluster ->
            ClusterMarker(
                cluster = cluster,
                onTap = { onClusterTap(cluster) },
                modifier = Modifier.offset(
                    x = (cluster.centerLon * 100).dp, // 실제 좌표 변환 필요
                    y = (cluster.centerLat * 100).dp
                )
            )
        }
    }
}

@Composable
private fun ClusterMarker(
    cluster: SignalCluster,
    onTap: () -> Unit,
    modifier: Modifier = Modifier
) {
    val scale by animateFloatAsState(
        targetValue = 1f,
        animationSpec = spring(dampingRatio = Spring.DampingRatioMediumBouncy)
    )

    Box(
        modifier = modifier
            .scale(scale)
            .size((40 + cluster.signalCount * 2).dp)
            .clip(CircleShape)
            .background(getClusterColor(cluster.signalCount))
            .border(2.dp, Color.White, CircleShape)
            .clickable { onTap() },
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = cluster.signalCount.toString(),
            color = Color.White,
            fontWeight = FontWeight.Bold,
            fontSize = 12.sp
        )
    }
}

private fun getClusterColor(count: Int): Color {
    return when {
        count < 5 -> Color.Blue
        count < 10 -> Color.Green
        count < 20 -> Color(0xFFFFA500) // Orange
        else -> Color.Red
    }
}

private fun calculateOpacity(density: Int, weight: Double): Float {
    val normalizedDensity = min(density / 10.0, 1.0)
    val normalizedWeight = min(weight / 5.0, 1.0)
    return max(normalizedDensity * normalizedWeight, 0.1).toFloat()
}

/**
 * 지도 레전드 위젯
 */
@Composable
fun MapLegend(
    viewMode: MapViewMode,
    densityPoints: List<DensityPoint>? = null,
    clusters: List<SignalCluster>? = null,
    modifier: Modifier = Modifier
) {
    AnimatedVisibility(
        visible = viewMode != MapViewMode.NORMAL,
        enter = slideInVertically() + fadeIn(),
        exit = slideOutVertically() + fadeOut(),
        modifier = modifier
    ) {
        Card(
            modifier = Modifier.padding(16.dp),
            shape = RoundedCornerShape(8.dp),
            elevation = CardDefaults.cardElevation(defaultElevation = 8.dp)
        ) {
            Column(
                modifier = Modifier.padding(12.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text(
                    text = getLegendTitle(viewMode),
                    fontWeight = FontWeight.Bold,
                    fontSize = 14.sp
                )

                when (viewMode) {
                    MapViewMode.DENSITY -> DensityLegendItems()
                    MapViewMode.CLUSTER -> ClusterLegendItems()
                    else -> {
                        Text(
                            text = "기본 마커 표시",
                            fontSize = 12.sp
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun DensityLegendItems() {
    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
        LegendItem(Color.Blue.copy(alpha = 0.3f), "낮음 (1-2개)")
        LegendItem(Color.Blue.copy(alpha = 0.6f), "보통 (3-5개)")
        LegendItem(Color.Blue.copy(alpha = 0.9f), "높음 (6개 이상)")
    }
}

@Composable
private fun ClusterLegendItems() {
    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
        LegendItem(Color.Blue, "소규모 (2-4개)")
        LegendItem(Color.Green, "중간 (5-9개)")
        LegendItem(Color(0xFFFFA500), "대규모 (10-19개)")
        LegendItem(Color.Red, "초대규모 (20개 이상)")
    }
}

@Composable
private fun LegendItem(
    color: Color,
    label: String
) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Box(
            modifier = Modifier
                .size(12.dp)
                .clip(CircleShape)
                .background(color)
        )
        Text(
            text = label,
            fontSize = 12.sp
        )
    }
}

private fun getLegendTitle(viewMode: MapViewMode): String {
    return when (viewMode) {
        MapViewMode.DENSITY -> "밀도 범례"
        MapViewMode.CLUSTER -> "클러스터 범례"
        else -> "일반 지도"
    }
}

/**
 * 지도 성능 모니터 위젯
 */
@Composable
fun MapPerformanceMonitor(
    signalCount: Int,
    clusterCount: Int,
    isLoading: Boolean,
    modifier: Modifier = Modifier
) {
    Card(
        modifier = modifier,
        colors = CardDefaults.cardColors(
            containerColor = Color.Black.copy(alpha = 0.7f),
            contentColor = Color.White
        ),
        shape = RoundedCornerShape(4.dp)
    ) {
        Column(
            modifier = Modifier.padding(8.dp),
            horizontalAlignment = Alignment.End,
            verticalArrangement = Arrangement.spacedBy(2.dp)
        ) {
            Text(
                text = "Signals: $signalCount",
                fontSize = 10.sp
            )

            if (clusterCount > 0) {
                Text(
                    text = "Clusters: $clusterCount",
                    fontSize = 10.sp
                )
            }

            if (isLoading) {
                CircularProgressIndicator(
                    modifier = Modifier.size(12.dp),
                    strokeWidth = 1.dp,
                    color = Color.White
                )
            }
        }
    }
}

File: ios/lib/features/signal/presentation/widgets/map_visualization_layer.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

import '../../data/services/enhanced_geographic_service.dart';
import '../cubit/enhanced_map_state.dart';

/// 지도 시각화 레이어 (밀도 맵, 클러스터, 핫스팟)
class MapVisualizationLayer {
  /// 밀도 맵을 위한 히트맵 오버레이 생성
  static Set<Polygon> createDensityHeatmap(
    List<DensityPoint> densityPoints, {
    double gridSize = 0.01,
    Color baseColor = Colors.blue,
  }) {
    final polygons = <Polygon>{};
    
    for (int i = 0; i < densityPoints.length; i++) {
      final point = densityPoints[i];
      final opacity = _calculateOpacity(point.density, point.weight);
      
      // 그리드 셀을 다각형으로 변환
      final polygon = Polygon(
        polygonId: PolygonId('density_$i'),
        points: _createGridCell(point.latitude, point.longitude, gridSize),
        fillColor: baseColor.withOpacity(opacity),
        strokeColor: baseColor.withOpacity(opacity * 0.7),
        strokeWidth: 1,
        consumeTapEvents: false,
      );
      
      polygons.add(polygon);
    }
    
    return polygons;
  }

  /// 클러스터 마커 생성
  static Set<Marker> createClusterMarkers(
    List<SignalCluster> clusters, {
    required Function(SignalCluster) onClusterTap,
  }) {
    final markers = <Marker>{};
    
    for (final cluster in clusters) {
      final marker = Marker(
        markerId: MarkerId('cluster_${cluster.id}'),
        position: LatLng(cluster.centerLat, cluster.centerLon),
        icon: _createClusterIcon(cluster.signalCount),
        onTap: () => onClusterTap(cluster),
        infoWindow: InfoWindow(
          title: '클러스터',
          snippet: '${cluster.signalCount}개 시그널',
        ),
      );
      
      markers.add(marker);
    }
    
    return markers;
  }

  /// 핫스팟 마커 생성
  static Set<Marker> createHotspotMarkers(
    List<Hotspot> hotspots, {
    required Function(Hotspot) onHotspotTap,
  }) {
    final markers = <Marker>{};
    
    for (final hotspot in hotspots) {
      final marker = Marker(
        markerId: MarkerId('hotspot_${hotspot.id}'),
        position: LatLng(hotspot.latitude, hotspot.longitude),
        icon: BitmapDescriptor.defaultMarkerWithHue(BitmapDescriptor.hueOrange),
        onTap: () => onHotspotTap(hotspot),
        infoWindow: InfoWindow(
          title: '핫스팟: ${hotspot.name}',
          snippet: '인기 지역',
        ),
      );
      
      markers.add(marker);
    }
    
    return markers;
  }

  /// 다각형 검색 영역 표시
  static Set<Polygon> createPolygonSearchArea(
    List<List<double>> polygonPoints, {
    Color fillColor = Colors.blue,
    Color strokeColor = Colors.blue,
  }) {
    if (polygonPoints.length < 3) return {};
    
    final points = polygonPoints
        .map((point) => LatLng(point[1], point[0])) // [lon, lat] -> LatLng
        .toList();
    
    return {
      Polygon(
        polygonId: const PolygonId('search_polygon'),
        points: points,
        fillColor: fillColor.withOpacity(0.2),
        strokeColor: strokeColor,
        strokeWidth: 2,
        consumeTapEvents: false,
      ),
    };
  }

  /// 경로 검색 영역 표시
  static Set<Polyline> createRouteSearchArea(
    double startLat,
    double startLon,
    double endLat,
    double endLon, {
    double bufferWidth = 1000,
    Color color = Colors.green,
  }) {
    return {
      Polyline(
        polylineId: const PolylineId('search_route'),
        points: [
          LatLng(startLat, startLon),
          LatLng(endLat, endLon),
        ],
        color: color,
        width: 5,
        patterns: [PatternItem.dash(20), PatternItem.gap(10)],
      ),
    };
  }

  /// 반경 검색 영역 표시
  static Set<Circle> createRadiusSearchArea(
    double centerLat,
    double centerLon,
    double radius, {
    Color fillColor = Colors.blue,
    Color strokeColor = Colors.blue,
  }) {
    return {
      Circle(
        circleId: const CircleId('search_radius'),
        center: LatLng(centerLat, centerLon),
        radius: radius,
        fillColor: fillColor.withOpacity(0.1),
        strokeColor: strokeColor,
        strokeWidth: 2,
      ),
    };
  }

  // Private Helper Methods

  static double _calculateOpacity(int density, double weight) {
    // 밀도와 가중치를 기반으로 투명도 계산
    final normalizedDensity = math.min(density / 10.0, 1.0);
    final normalizedWeight = math.min(weight / 5.0, 1.0);
    return math.max(normalizedDensity * normalizedWeight, 0.1);
  }

  static List<LatLng> _createGridCell(
    double centerLat,
    double centerLon,
    double gridSize,
  ) {
    final halfGrid = gridSize / 2;
    return [
      LatLng(centerLat - halfGrid, centerLon - halfGrid),
      LatLng(centerLat - halfGrid, centerLon + halfGrid),
      LatLng(centerLat + halfGrid, centerLon + halfGrid),
      LatLng(centerLat + halfGrid, centerLon - halfGrid),
    ];
  }

  static BitmapDescriptor _createClusterIcon(int count) {
    // 클러스터 크기에 따른 아이콘 선택
    if (count < 5) {
      return BitmapDescriptor.defaultMarkerWithHue(BitmapDescriptor.hueBlue);
    } else if (count < 10) {
      return BitmapDescriptor.defaultMarkerWithHue(BitmapDescriptor.hueGreen);
    } else if (count < 20) {
      return BitmapDescriptor.defaultMarkerWithHue(BitmapDescriptor.hueYellow);
    } else {
      return BitmapDescriptor.defaultMarkerWithHue(BitmapDescriptor.hueRed);
    }
  }
}

/// 실시간 애니메이션을 위한 커스텀 위젯
class AnimatedMapOverlay extends StatefulWidget {
  final List<DensityPoint> densityPoints;
  final List<SignalCluster> clusters;
  final MapViewMode viewMode;
  final Function(SignalCluster) onClusterTap;

  const AnimatedMapOverlay({
    super.key,
    required this.densityPoints,
    required this.clusters,
    required this.viewMode,
    required this.onClusterTap,
  });

  @override
  State<AnimatedMapOverlay> createState() => _AnimatedMapOverlayState();
}

class _AnimatedMapOverlayState extends State<AnimatedMapOverlay>
    with TickerProviderStateMixin {
  late AnimationController _pulseController;
  late AnimationController _fadeController;

  @override
  void initState() {
    super.initState();
    _pulseController = AnimationController(
      duration: const Duration(seconds: 2),
      vsync: this,
    )..repeat();

    _fadeController = AnimationController(
      duration: const Duration(milliseconds: 500),
      vsync: this,
    );
  }

  @override
  void didUpdateWidget(AnimatedMapOverlay oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.viewMode != widget.viewMode) {
      _fadeController.forward();
    }
  }

  @override
  void dispose() {
    _pulseController.dispose();
    _fadeController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: Listenable.merge([_pulseController, _fadeController]),
      builder: (context, child) {
        return Opacity(
          opacity: _fadeController.value,
          child: _buildOverlayContent(),
        );
      },
    );
  }

  Widget _buildOverlayContent() {
    switch (widget.viewMode) {
      case MapViewMode.density:
        return _buildDensityOverlay();
      case MapViewMode.cluster:
        return _buildClusterOverlay();
      case MapViewMode.normal:
      default:
        return const SizedBox.shrink();
    }
  }

  Widget _buildDensityOverlay() {
    return Stack(
      children: widget.densityPoints.map((point) {
        return Positioned(
          left: point.longitude * 100, // 임시 좌표 변환
          top: point.latitude * 100,
          child: Container(
            width: 20,
            height: 20,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.red.withOpacity(
                0.3 + (0.7 * _pulseController.value),
              ),
            ),
          ),
        );
      }).toList(),
    );
  }

  Widget _buildClusterOverlay() {
    return Stack(
      children: widget.clusters.map((cluster) {
        return Positioned(
          left: cluster.centerLon * 100, // 임시 좌표 변환
          top: cluster.centerLat * 100,
          child: GestureDetector(
            onTap: () => widget.onClusterTap(cluster),
            child: AnimatedContainer(
              duration: const Duration(milliseconds: 300),
              width: 40 + (cluster.signalCount * 2),
              height: 40 + (cluster.signalCount * 2),
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                color: _getClusterColor(cluster.signalCount),
                border: Border.all(
                  color: Colors.white,
                  width: 2,
                ),
                boxShadow: [
                  BoxShadow(
                    color: Colors.black.withOpacity(0.3),
                    blurRadius: 6,
                    offset: const Offset(0, 2),
                  ),
                ],
              ),
              child: Center(
                child: Text(
                  '${cluster.signalCount}',
                  style: const TextStyle(
                    color: Colors.white,
                    fontWeight: FontWeight.bold,
                    fontSize: 12,
                  ),
                ),
              ),
            ),
          ),
        );
      }).toList(),
    );
  }

  Color _getClusterColor(int count) {
    if (count < 5) {
      return Colors.blue;
    } else if (count < 10) {
      return Colors.green;
    } else if (count < 20) {
      return Colors.orange;
    } else {
      return Colors.red;
    }
  }
}

/// 지도 레전드 위젯
class MapLegend extends StatelessWidget {
  final MapViewMode viewMode;
  final List<DensityPoint>? densityPoints;
  final List<SignalCluster>? clusters;

  const MapLegend({
    super.key,
    required this.viewMode,
    this.densityPoints,
    this.clusters,
  });

  @override
  Widget build(BuildContext context) {
    return Positioned(
      bottom: 100,
      left: 16,
      child: Container(
        padding: const EdgeInsets.all(12),
        decoration: BoxDecoration(
          color: Colors.white,
          borderRadius: BorderRadius.circular(8),
          boxShadow: [
            BoxShadow(
              color: Colors.black.withOpacity(0.1),
              blurRadius: 8,
              offset: const Offset(0, 2),
            ),
          ],
        ),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          mainAxisSize: MainAxisSize.min,
          children: [
            Text(
              _getLegendTitle(),
              style: const TextStyle(
                fontWeight: FontWeight.bold,
                fontSize: 14,
              ),
            ),
            const SizedBox(height: 8),
            ..._buildLegendItems(),
          ],
        ),
      ),
    );
  }

  String _getLegendTitle() {
    switch (viewMode) {
      case MapViewMode.density:
        return '밀도 범례';
      case MapViewMode.cluster:
        return '클러스터 범례';
      case MapViewMode.normal:
      default:
        return '일반 지도';
    }
  }

  List<Widget> _buildLegendItems() {
    switch (viewMode) {
      case MapViewMode.density:
        return _buildDensityLegend();
      case MapViewMode.cluster:
        return _buildClusterLegend();
      case MapViewMode.normal:
      default:
        return [const Text('기본 마커 표시')];
    }
  }

  List<Widget> _buildDensityLegend() {
    return [
      _buildLegendItem(Colors.blue.withOpacity(0.3), '낮음 (1-2개)'),
      _buildLegendItem(Colors.blue.withOpacity(0.6), '보통 (3-5개)'),
      _buildLegendItem(Colors.blue.withOpacity(0.9), '높음 (6개 이상)'),
    ];
  }

  List<Widget> _buildClusterLegend() {
    return [
      _buildLegendItem(Colors.blue, '소규모 (2-4개)'),
      _buildLegendItem(Colors.green, '중간 (5-9개)'),
      _buildLegendItem(Colors.orange, '대규모 (10-19개)'),
      _buildLegendItem(Colors.red, '초대규모 (20개 이상)'),
    ];
  }

  Widget _buildLegendItem(Color color, String label) {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 2),
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          Container(
            width: 12,
            height: 12,
            decoration: BoxDecoration(
              color: color,
              shape: BoxShape.circle,
            ),
          ),
          const SizedBox(width: 8),
          Text(
            label,
            style: const TextStyle(fontSize: 12),
          ),
        ],
      ),
    );
  }
}

/// 지도 성능 모니터 위젯
class MapPerformanceMonitor extends StatefulWidget {
  final int signalCount;
  final int clusterCount;
  final bool isLoading;

  const MapPerformanceMonitor({
    super.key,
    required this.signalCount,
    required this.clusterCount,
    required this.isLoading,
  });

  @override
  State<MapPerformanceMonitor> createState() => _MapPerformanceMonitorState();
}

class _MapPerformanceMonitorState extends State<MapPerformanceMonitor> {
  @override
  Widget build(BuildContext context) {
    return Positioned(
      top: 100,
      right: 16,
      child: Container(
        padding: const EdgeInsets.all(8),
        decoration: BoxDecoration(
          color: Colors.black.withOpacity(0.7),
          borderRadius: BorderRadius.circular(4),
        ),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.end,
          mainAxisSize: MainAxisSize.min,
          children: [
            Text(
              'Signals: ${widget.signalCount}',
              style: const TextStyle(
                color: Colors.white,
                fontSize: 10,
              ),
            ),
            if (widget.clusterCount > 0)
              Text(
                'Clusters: ${widget.clusterCount}',
                style: const TextStyle(
                  color: Colors.white,
                  fontSize: 10,
                ),
              ),
            if (widget.isLoading)
              const SizedBox(
                width: 12,
                height: 12,
                child: CircularProgressIndicator(
                  strokeWidth: 1,
                  valueColor: AlwaysStoppedAnimation<Color>(Colors.white),
                ),
              ),
          ],
        ),
      ),
    );
  }
}